src/jwtserviceauth.js

```diff
@@ -174,6 +138,98 @@ class JwtServiceAuth {
   }
 }
 
+/**
+ * Get Google Access Token from service account keyfile
+ * @param {string} keyFileData
+ * @param {Array<string>} [scopes=['https://www.googleapis.com/auth/userinfo.email']]
+ * @param {Object} [options]
+ * @param {number} [options.expires=3600]
+ * @param {string} [options.impersonate]
+ * @returns {Promise<accessTokenResponse>}
+ */
+function _getGoogleAccessToken(
+  httpRequestHandler,
+  keyFileData,
+  scopes,
+  options
+) {
+  // TODO: Remove in V2.0
+  // Support old interface for expires
+  if (typeof scopes === 'number') {
+    if (options !== null && typeof options === 'object') {
+      options.expires = scopes
+    } else {
+      options = {
+        expires: scopes
+      }
+    }
+    scopes = null
+  }
+
+  scopes = scopes ? scopes : ['https://www.googleapis.com/auth/userinfo.email']
+
+  let keyData = JSON.parse(keyFileData)
+
+  if (keyData.type !== 'service_account') {
+    throw new Error('Only supports service account keyFiles')
+  }
+
+  // Create JWT auth token
+  let unixNow = Math.floor(new Date().getTime() / 1000)
+  let jwtHeader = {
+    typ: 'JWT',
+    alg: 'RS256',
+    kid: keyData.private_key_id
+  }
+  let jwtBody = {
+    aud: 'https://www.googleapis.com/oauth2/v4/token',
+    iss: keyData.client_email,
+    iat: unixNow,
+    exp: unixNow + (options.expires || 3600),
+    scope: scopes.join(' ')
+  }
+
+  if (options.impersonate) {
+    jwtBody.sub = options.impersonate
+  }
+
+  let jwt = jwtEncode(keyData.private_key, jwtHeader, jwtBody)
+
+  let formParams = {
+    grant_type: 'urn:ietf:params:oauth:grant-type:jwt-bearer',
+    assertion: jwt
+  }
+  let formData = Object.keys(formParams)
+    .map(key => `${key}=${querystring.escape(formParams[key])}`)
+    .join('&')
+
+  // Fetch access token
+  return httpRequestHandler(
+    'POST',
+    `https://www.googleapis.com/oauth2/v4/token`,
+    {
+      'Content-Type': 'application/x-www-form-urlencoded',
+      'User-Agent': 'tlbdk-buildstatus'
+    },
+    formData
+  ).then(response => {
+    if (response.statusCode === 200) {
+      let authResponse = JSON.parse(Buffer.from(response.data).toString('utf8'))
+      let now = new Date().getTime()
```

Comment:
move the time stamp to the previous scope to avoid a racecondition where you think expiresAt is valid, but its invalid on the other end in reality, due to io delays